zone/cheat_manager.cpp

```diff
@@ -0,0 +1,385 @@
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName()
+				);
+				LogCheat(message);
+			}
+			break;
+
+		case MQWarpLight:
+			if (RuleB(Zone, EnableMQWarpDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQWarpExemptStatus) || (RuleI(Zone, MQWarpExemptStatus)) == -1))) {
+				if (RuleB(Zone, MarkMQWarpLT)) {
+					std::string message = fmt::format(
+						"/MQWarp(Knockback) with location from x [{:.2f}] y [{:.2f}] z [{:.2f}] running fast but not fast enough to get killed, possibly: small warp, speed hack, excessive lag, marked as suspicious.",
+						position1.x,
+						position1.y,
+						position1.z
+					);
+					database.SetMQDetectionFlag(
+						m_target->AccountName(),
+						m_target->GetName(),
+						message.c_str(),
+						zone->GetShortName()
+					);
+					LogCheat(message);
+				}
+			}
+			break;
+
+		case MQZone:
+			if (RuleB(Zone, EnableMQZoneDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQZoneExemptStatus) || (RuleI(Zone, MQZoneExemptStatus)) == -1))) {
+				std::string message = fmt::format(
+					"/MQZone used at x [{:.2f}] y [{:.2f}] z [{:.2f}]",
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName()
+				);
+				LogCheat(message);
+			}
+			break;
+		case MQZoneUnknownDest:
+			if (RuleB(Zone, EnableMQZoneDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQZoneExemptStatus) || (RuleI(Zone, MQZoneExemptStatus)) == -1))) {
+				std::string message = fmt::format(
+					"/MQZone used at x [{:.2f}] y [{:.2f}] z [{:.2f}] with Unknown Destination",
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName());
+				LogCheat(message.c_str());
+			}
+			break;
+		case MQGate:
+			if (RuleB(Zone, EnableMQGateDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQGateExemptStatus) || (RuleI(Zone, MQGateExemptStatus)) == -1))) {
+				std::string message = fmt::format(
+					"/MQGate used at x [{:.2f}] y [{:.2f}] z [{:.2f}]",
+					position1.x,
+					position1.y,
+					position1.z
+				);
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					message.c_str(),
+					zone->GetShortName()
+				);
+				LogCheat(message);
+			}
+			break;
+		case MQGhost:
+			if (RuleB(Zone, EnableMQGhostDetector) &&
+				((m_target->Admin() < RuleI(Zone, MQGhostExemptStatus) || (RuleI(Zone, MQGhostExemptStatus)) == -1))) {
+				database.SetMQDetectionFlag(
+					m_target->AccountName(),
+					m_target->GetName(),
+					"/MQGhost",
+					zone->GetShortName());
+				LogCheat("/MQGhost");
```

Comment:
This doesn't do much or tell anyone anything

Reply:
its something that needs to be better developed, mqghost is when MQ2 blocks the ClientUpdate packets so you can move freely around the zone client side but since it doesn't update server side you won't aggro NPCs... differnt builds of MQ2 have this differently implemented.